kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -162,7 +163,9 @@ public Transaction build(Engine engine) {
       newMetadata = Optional.of(snapshotMetadata.withMergedConfiguration(newProperties));
     }
 
-    // TODO In the future update metadata with new schema if provided
+    if (schema.isPresent() && !isNewTable) {
+      newMetadata = Optional.of(newMetadata.orElse(snapshotMetadata).withNewSchema(schema.get()));
```

Comment:
shouldn't the new schema validations be here?

Reply:
I intentionally do it after the protocol is updated based on table properties, in step 2 below. For instance if someone enables icebergCompatV2, as part of the transaction we want the validation to be applied on the new metadata. We do validation after all the metadata has been updated.

Reply:
The problem is other validations between this step and the validations assume certain schema elements are correct (e.g. icebergCompatV2 NPE that was handled in this PR). I think splitting along the logical validation here and physical validation at the end seems like the right split.

I am ok to keep this like now, if we encounter more cases like the icebergCompatV2, we can revisit the code here.